Add support for running a single test

younata opened the pull request:
This implements half of the functionality of [zacwest's PR](https://github.com/Quick/Quick/pull/936), but in a more maintainable manner (this implementation uses `XCTTestIdentifier -components`, whereas zacwest's tries to parse the string from `_identifierString`, which doesn't work as well because `_identifierString` can return either swift-formatted (`foo/myMethod()`) or objc-formatted (`foo/myMethod`) identifiers).
This allows you to select an individual (or set of) test from the Test Navigator in Xcode and run only that test.

This also adds tests guaranteeing the functionality works, though it is split between tests written in objc and swift.

The PR should summarize what was changed and why. Here are some questions to
help you if you're not sure:

 - What behavior was changed?
     - This adds the ability to run an individual test. You can do this in Xcode using the Test Navigator. Eventually (not in this PR), I'd love to figure out a way to have the "test diamonds" appear next to all the "it" tests.
 - What code was refactored / updated to support this change?
    - Moved logic for determining name of test case class to run from `QuickSelectedTestSuiteBuilder testCaseClassForTestCaseWithName` to `XCTestSuite+qck_hooked_testSuiteForTestCaseWithName:` (the part where we do the `[name componentsSeparatedByString:@"/"]`). This was done to support pulling out the individual test name using mechanisms that XCTest gives us.
     - Instead of parsing the `XCTTestIdentifier._identifierString` property to figure out which test case to run, we use the `components` property of `XCTTestIdentifier`, which gives us the test case class to run (`components.firstObject`), then the individual test to run (`components.lastObject`, there are only ever 2 objects in the `components` array).
     - Changed `QuickTestSuite.selectedTestSuite(forTestCaseWithName:)` to `QuickTestSuite.selectedTestSuite(forTestCaseWithName:testName:)`, which takes an optional `testName` (for the individual test to run) and constructs the objective-c style name of the test, to later filter down for the exact test to run.

Checklist - While not every PR needs it, new features should consider this list:

 - [x] Does this have tests?
 - [x] Does this have documentation?
 - Does not break the public API.
 - [x] Is this a new feature (Requires minor version bump)?

younata commented:
@ikesyo, any thoughts on this?

NachoSoto commented:
This would be awesome, I've always missed this feature from Quick.

jessesquires commented:
I'd like to include this in the next release. I just rebased the branch and trigger GH actions to run.

jessesquires commented:
Hey @younata! 👋🏼 😄 Are you (or anyone else on this thread) able to see this PR through? 

I would love to get it merged soon for v5. Looks like there are a number of CI failures though.

younata commented:
> Hey @younata! 👋🏼 😄 Are you (or anyone else on this thread) able to see this PR through?
> 
> I would love to get it merged soon for v5. Looks like there are a number of CI failures though.

I'm working on addressing those CI failures. It looks like the 12.4 test failures can be resolved by skipping those specific tests when run on Xcode 12.4 (which do call out that the API it uses is only available on Xcode 12.5 and above).

jessesquires commented:
Thanks @younata ! 😄

younata commented:
Hey @jessesquires, thanks so much for your help! I rebased on top of main, and added a commit to remove one last build warning. Once those go through CI, I'll merge this.

jessesquires commented:
@younata sounds good! Thanks for the contribution! 😄

younata commented:
@jessesquires Actually, I don't have merge privy (of course). Would you mind merging it?

jessesquires commented:
@younata I invited you to be a maintainer 😄 

Thanks again for following up on this! 💯

NachoSoto commented on Tests/QuickTests/QuickTests/FunctionalTests/ObjC/SelectedTests+ObjC.m:
```diff
@@ -0,0 +1,93 @@
+
+it(@"correctly grabs all tests in a test case if no specific test is specified", ^{
+    if ([XCTestSuite respondsToSelector:@selector(testClassSuitesForTestIdentifiers:skippingTestIdentifiers:randomNumberGenerator:)] == NO) {
+        XCTSkip(@"Skipping. XCTestSuite does not respond to testClassSuitesForTestIdentifiers:skippingTestIdentifiers:randomNumberGenerator:");
```
You can use `XCTSkipIf`

Blackjacx commented:
![Screen Shot 2022-08-23 at 11 01 01](https://user-images.githubusercontent.com/794372/186117697-f013e240-8af4-4205-a641-da979a3e11bb.png)


Offtopic: How can I have this in my PRs?!
